### ISSUE TITLE
Depreciation warning

### ISSUE THREAD

**Issue by yoann-antoviaque** [hanscappelle/zero-motorcycles-integration#52]
## System Information

version | core-2024.7.3
-- | --
installation_type | Home Assistant OS
dev | false
hassio | true
docker | true
user | root
virtualenv | false
python_version | 3.12.4
os_name | Linux
os_version | 6.6.31-haos-raspi
arch | aarch64
timezone | Europe/Paris
config_dir | /config

<details><summary>Home Assistant Community Store</summary>

GitHub API | ok
-- | --
GitHub Content | ok
GitHub Web | ok
GitHub API Calls Remaining | 5000
Installed Version | 1.34.0
Stage | running
Available Repositories | 1391
Downloaded Repositories | 3
HACS Data | ok

</details>

<details><summary>Home Assistant Cloud</summary>

logged_in | false
-- | --
can_reach_cert_server | ok
can_reach_cloud_auth | ok
can_reach_cloud | ok

</details>

<details><summary>Home Assistant Supervisor</summary>

host_os | Home Assistant OS 12.4
-- | --
update_channel | stable
supervisor_version | supervisor-2024.06.2
agent_version | 1.6.0
docker_version | 26.1.4
disk_total | 57.8 GB
disk_used | 4.6 GB
healthy | true
supported | true
host_connectivity | true
supervisor_connectivity | true
ntp_synchronized | true
virtualization | 
board | rpi3-64
supervisor_api | ok
version_api | ok
installed_addons | File editor (5.8.0), Advanced SSH & Web Terminal (18.0.0), Studio Code Server (5.15.0)

</details>

<details><summary>Dashboards</summary>

dashboards | 2
-- | --
resources | 0
views | 0
mode | storage

</details>

<details><summary>Recorder</summary>

oldest_recorder_run | 26 juillet 2024 à 09:38
-- | --
current_recorder_run | 31 juillet 2024 à 14:40
estimated_db_size | 14.61 MiB
database_engine | sqlite
database_version | 3.45.3

</details>

### Checklist

- [X] I have enabled debug logging for my installation.
- [X] I have filled out the issue template to the best of my ability.
- [X] This issue only contains 1 issue (if you have multiple issues, open one issue for each issue).
- [X] This issue is not a duplicate issue of currently [previous issues](https://github.com/hanscappelle/zero_motorcycles_integration/issues?q=is%3Aissue+label%3A%22Bug%22+)..

### Describe the issue

The system warns about a deprecated constant that will be removed, it does not block anything yet, but thought it would be nice to report it before it does

> SOURCE_TYPE_GPS was used from zero_motorcycles_integration, this is a deprecated constant which will be removed in HA Core 2025.1. Use SourceType.GPS instead, please report it to the author of the 'zero_motorcycles_integration' custom integration

### Reproduction steps

1. add zero-motorcycles-integration
2. reboot HA
3. look at the logs

### Debug logs

```text
2024-07-31 14:40:36.604 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration vaillant_vsmart which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2024-07-31 14:40:36.610 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration zero_motorcycles_integration which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2024-07-31 14:40:36.615 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration hacs which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2024-07-31 14:40:40.015 WARNING (Recorder) [homeassistant.components.recorder.util] The system could not validate that the sqlite3 database at //config/home-assistant_v2.db was shutdown cleanly
2024-07-31 14:40:40.343 WARNING (Recorder) [homeassistant.components.recorder.util] Ended unfinished session (id=10 from 2024-07-28 15:23:50.135012)
2024-07-31 14:40:46.742 ERROR (MainThread) [homeassistant.config] Invalid config for 'tts' from integration 'google_translate' at configuration.yaml, line 36: 'base_url' is an invalid option for 'google_translate.tts', check: base_url, please check the docs at https://www.home-assistant.io/integrations/google_translate
2024-07-31 14:40:51.879 WARNING (MainThread) [bluetooth_auto_recovery.recover] Could not determine the power state of the Bluetooth adapter hci0 [00:00:00:00:00:00] due to timeout after 5 seconds
2024-07-31 14:41:12.916 WARNING (MainThread) [bluetooth_auto_recovery.recover] Could not cycle the Bluetooth adapter hci0 [00:00:00:00:00:00]: [Errno 110] Operation timed out
2024-07-31 14:41:12.920 WARNING (MainThread) [bluetooth_auto_recovery.recover] Bluetooth management socket connection lost: [Errno 22] Invalid argument
2024-07-31 14:41:17.919 WARNING (MainThread) [bluetooth_auto_recovery.recover] Could not reset the power state of the Bluetooth adapter hci0 [00:00:00:00:00:00] due to timeout after 5 seconds
2024-07-31 14:41:17.925 WARNING (MainThread) [bluetooth_auto_recovery.recover] Closing Bluetooth adapter hci0 [00:00:00:00:00:00] failed: [Errno 9] Bad file descriptor
2024-07-31 14:41:26.650 WARNING (MainThread) [homeassistant.helpers.frame] Detected that custom integration 'hacs' calls hass.http.register_static_path which is deprecated because it does blocking I/O in the event loop, instead call `await hass.http.async_register_static_paths([StaticPathConfig("/hacsfiles/frontend", "/config/custom_components/hacs/hacs_frontend_experimental", False)])`; This function will be removed in 2025.7 at custom_components/hacs/frontend.py, line 49: hass.http.register_static_path(, please create a bug report at https://github.com/hacs/integration/issues
2024-07-31 14:41:26.667 WARNING (MainThread) [homeassistant.helpers.frame] Detected that custom integration 'hacs' calls hass.http.register_static_path which is deprecated because it does blocking I/O in the event loop, instead call `await hass.http.async_register_static_paths([StaticPathConfig("/hacsfiles/iconset.js", "/config/custom_components/hacs/iconset.js", True)])`; This function will be removed in 2025.7 at custom_components/hacs/frontend.py, line 57: hass.http.register_static_path(, please create a bug report at https://github.com/hacs/integration/issues
2024-07-31 14:41:26.686 WARNING (MainThread) [homeassistant.helpers.frame] Detected that custom integration 'hacs' accesses hass.components.frontend. This is deprecated and will stop working in Home Assistant 2024.9, it should be updated to import functions used from frontend directly at custom_components/hacs/frontend.py, line 68: hass.components.frontend.async_register_built_in_panel(, please create a bug report at https://github.com/hacs/integration/issues
2024-07-31 14:41:41.808 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: {('shopping_list', '01J3Q6SC6F28SBZN6193176MM5'): 144.51724526, ('hacs', '01J3SZ1B21STJSMXY87EST9N0B'): 148.148225357, ('androidtv_remote', '01J3Q6TWNAVEJ1YHF3XEGCMMP3'): 150.7391511, ('ipp', '01J3Q796RRZCB31HG1M853T9KF'): 151.004406827, ('broadlink', '01J3RNDB65QB2WWSE0M6SCXBW8'): 151.218702425, ('doorbird', '01J3R21Q1NTJTQCCT7JY23G0PS'): 151.463262101, ('zero_motorcycles_integration', '01J3T0ZVW538ZACDFBSK1K3Z2Q'): 151.525341632, ('vaillant_vsmart', '01J3WHZDZJKV1KDVER6YGJJG60'): 154.91791151, ('synology_dsm', '01J3R2V8SEVEVP5P7YJ3EH8BW4'): 155.300939717, ('rpi_power', '01J3Q6HQWMC5NZQ0109CDRPSYD'): 155.509659221, ('cast', '01J3Q6HSD6PGJGXX3RNAG6MVAA'): 155.97174927, ('met', '01J3Q6SCCWE12ATDNDRCBB244B'): 156.124621193}
2024-07-31 14:41:46.740 WARNING (ImportExecutor_0) [homeassistant.components.device_tracker] SOURCE_TYPE_GPS was used from zero_motorcycles_integration, this is a deprecated constant which will be removed in HA Core 2025.1. Use SourceType.GPS instead, please report it to the author of the 'zero_motorcycles_integration' custom integration
2024-07-31 14:41:46.761 WARNING (ImportExecutor_0) [homeassistant.components.device_tracker] SOURCE_TYPE_GPS was used from zero_motorcycles_integration, this is a deprecated constant which will be removed in HA Core 2025.1. Use SourceType.GPS instead, please report it to the author of the 'zero_motorcycles_integration' custom integration
```

**Comment by stephenmhall** [comment]
This has now come to pass and the integration has stopped working

**Comment by ChrisBGL** [comment]
Please fix it 🙏

**Comment by hanscappelle** [comment]
I've created this release based on input from the PR. Let me know if that resolves the problem or not. 

https://github.com/hanscappelle/zero-motorcycles-integration/releases/tag/0.5.0

**Comment by AdamCogswell** [comment]
Hey Hans! I refreshed my custom repo in HACS to pull version 0.5.0 and it did download the updated files. After restarting HASS, I manually deleted my _existing_ integration, and then re-added it. It now loads the integration successfully, but there are no devices and hence no entities associated with it.  Based on the log file, looks like it might be an issue with the device_tracker.py around the unit of measure. My latest log entry below:

```
2025-01-09 05:01:45.189 ERROR (MainThread) [homeassistant.components.device_tracker] Error while setting up zero_motorcycles_integration platform for device_tracker
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 366, in _async_setup_platform
    await asyncio.shield(awaitable)
  File "/config/custom_components/zero_motorcycles_integration/device_tracker.py", line 17, in async_setup_entry
    ZeroTrackerEntity(
    ~~~~~~~~~~~~~~~~~^
        coordinator=coordinator,
        ^^^^^^^^^^^^^^^^^^^^^^^^
        unitnumber=unit["unitnumber"],
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/config/custom_components/zero_motorcycles_integration/device_tracker.py", line 34, in __init__
    super().__init__(coordinator)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "/config/custom_components/zero_motorcycles_integration/entity.py", line 29, in __init__
    self._attr_attribution = info["attribution"]
                             ~~~~^^^^^^^^^^^^^^^
KeyError: 'attribution'
```

I haven't really dug into this further yet as my day job is getting in the way, but I'm going to try and take a deeper dive later today. Thanks for the work on this.

**Comment by hanscappelle** [comment]
I don't have any credentials anymore to test E2E but I do find a similar issue when updating on my setup. The trace is different:

```
Logger: homeassistant.components.sensor
Source: helpers/entity_platform.py:356
integration: Sensor ([documentation](https://www.home-assistant.io/integrations/sensor), [issues](https://github.com/home-assistant/core/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+sensor%22))
First occurred: 18:31:58 (1 occurrences)
Last logged: 18:31:58
Error while setting up rest platform for sensor

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 356, in _async_setup_platform
    await asyncio.shield(awaitable)
  File "/usr/src/homeassistant/homeassistant/components/rest/sensor.py", line 78, in async_setup_platform
    conf, coordinator, rest = await async_get_config_and_coordinator(
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/rest/__init__.py", line 150, in async_get_config_and_coordinator
    shared_data = hass.data[DOMAIN][REST_DATA][discovery_info[REST_IDX]]
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
IndexError: list index out of range
```

**Comment by hanscappelle** [comment]
It's been 2 years since I checked this code :o. Looks like there was an update in the sensor config as is handled here by iterating all sensors but for some it's incomplete. Index out of bounds error

```
for unit in coordinator.units:
        for sensor in SENSORS:
            devices.append(
                ZeroBinarySensor(
                    coordinator=coordinator,
                    entity_description=sensor,
                    unitnumber=unit["unitnumber"],
                    sensor_name=sensor.name,
                )
            )
```

**Comment by hanscappelle** [comment]
I see, some are missing config parameters

```
SENSORS = (
    SensorEntityDescription(
        key="zero_motorcycles",
        name="soc",
        icon="mdi:battery-charging-50",
        device_class=SensorDeviceClass.BATTERY,
        native_unit_of_measurement=PERCENTAGE,
    ),
    SensorEntityDescription(
        key="zero_motorcycles",
        name="name",
        icon="mdi:id-card",
    ),
    SensorEntityDescription(
        key="zero_motorcycles",
        name="mileage",
        icon="mdi:gauge",
        device_class=SensorDeviceClass.DISTANCE,
        native_unit_of_measurement=UnitOfLength.KILOMETERS,
    ),
```

**Comment by hanscappelle** [comment]
I'll push version `0.5.1` with a fix for the sensor config. Since I don't have credentials the best I can do is test adding it and then I do get to this auth error.

```
Logger: custom_components.zero_motorcycles_integration
Source: custom_components/zero_motorcycles_integration/config_flow.py:32
integration: zero_motorcycles_integration ([documentation](https://github.com/hanscappelle/zero-motorcycles-integration), [issues](https://github.com/hanscappelle/zero-motorcycles-integration/issues))
First occurred: 19:01:36 (1 occurrences)
Last logged: 19:01:36
Something really wrong happened!

Traceback (most recent call last):
  File "/config/custom_components/zero_motorcycles_integration/api.py", line 81, in _api_wrapper
    raise ZeroApiClientAuthenticationError(
custom_components.zero_motorcycles_integration.api.ZeroApiClientAuthenticationError: Invalid credentials

The above exception was the direct cause of the following exception:
```

Let me know if someone with credentials can continue this test for me. Thanks!

**Comment by AdamCogswell** [comment]
I'll test it after push. 😊

**Comment by AdamCogswell** [comment]
@hanscappelle - still getting an error but since this issue was scoped to the GPS deprecation warning (which IS fixed), I would suggest we close this issue and start a new one related to the errors currently happening. I'm happy to log a new issue. Thanks.

**Comment by stephenmhall** [comment]
I'm getting 3 errors, setting up device tracker, binary sensor and sensor all look like this

`Logger: homeassistant.components.sensor
Source: helpers/entity_platform.py:366
integration: Sensor (documentation, issues)
First occurred: 17:04:24 (1 occurrences)
Last logged: 17:04:24

Error while setting up zero_motorcycles_integration platform for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 366, in _async_setup_platform
    await asyncio.shield(awaitable)
  File "/config/custom_components/zero_motorcycles_integration/sensor.py", line 114, in async_setup_entry
    ZeroSensor(
    ~~~~~~~~~~^
        coordinator=coordinator,
        ^^^^^^^^^^^^^^^^^^^^^^^^
    ...<2 lines>...
        sensor_name=sensor.name,
        ^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/config/custom_components/zero_motorcycles_integration/sensor.py", line 136, in __init__
    super().__init__(coordinator)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "/config/custom_components/zero_motorcycles_integration/entity.py", line 29, in __init__
    self._attr_attribution = info["attribution"]
                             ~~~~^^^^^^^^^^^^^^^
KeyError: 'attribution'`

**Comment by sordoneF** [comment]
Same errors on my end, no entities.

**Comment by hanscappelle** [comment]
Okay found a fix for the `KeyError: 'attribution'` issue, resolved it locally, pushing changes now and will create version `0.5.2`

**Comment by hanscappelle** [comment]
done, keep me posted

**Comment by pwtanagit** [comment]
updated to 0.5.2 : well working, all entities appear

@hanscappelle  thanks  :-)

**Comment by AdamCogswell** [comment]
Can confirm. Working after 0.5.2 update.

**Comment by hanscappelle** [comment]
Great news. Thanks for the updates. Closing this ticket.
